cloudsync/youtube.py

```diff
@@ -0,0 +1,260 @@
+""" YouTube API interface"""
+import http
+import time
+import logging
+from tempfile import NamedTemporaryFile
+
+import boto3
+import httplib2
+import oauth2client
+
+from django.conf import settings
+from googleapiclient.discovery import build
+from googleapiclient.errors import HttpError
+from googleapiclient.http import MediaFileUpload
+
+from ui.encodings import EncodingNames
+
+log = logging.getLogger(__name__)
+
+
+def resumable_upload(request):
+    """
+    Upload a video to YouTube and resume on failure up to 10 times, adapted from YouTube API example
+
+    Args:
+        request(googleapiclient.http.HttpRequest): The Youtube API execute request to process
+
+    Returns:
+        dict: The YouTube API response
+    """
+    response = None
+    error = None
+    error_msg = None
+
+    retry = 0
+    max_retries = 10
+
+    retry_exceptions = (IOError, httplib2.HttpLib2Error, http.client.NotConnected,
+                        http.client.IncompleteRead, http.client.ImproperConnectionState,
+                        http.client.CannotSendRequest, http.client.CannotSendHeader,
+                        http.client.ResponseNotReady, http.client.BadStatusLine)
+    retry_statuses = [500, 502, 503, 504]
+
+    while response is None:
+        try:
+            _, response = request.next_chunk()
+            if response is None or 'id' not in response:
+                raise Exception('YouTube upload failed: %s' % response)
+        except HttpError as e:
+            if e.resp.status in retry_statuses:
+                error = e
+                error_msg = 'A retriable error %d occurred:\n%s' % (e.resp.status, e.content)
+            else:
+                raise
+        except retry_exceptions as e:
+            error = e
+            error_msg = 'A retriable error occurred: %s' % str(e)
+
+        if error is not None:
+            retry += 1
+            if retry > max_retries:
+                log.error('Retried YouTube upload 10x, giving up. Last error: %s', error_msg)
+                raise Exception(e)
+            max_sleep = 2 ** retry
```

Comment:
Can you rename to `sleep_time`? I'm not sure if `max` is appropriate here since there's no range

Reply:
OK